internal/services/executor/executor.go

```diff
@@ -1517,3 +1523,51 @@ func (e *Executor) Run(ctx context.Context) error {
 
 	return nil
 }
+
+func (e *Executor) tasksTimeoutCleanerLoop(ctx context.Context) {
+	for {
+		e.log.Debug().Msgf("tasksTimeoutCleaner")
+
+		e.tasksTimeoutCleaner(ctx)
+
+		sleepCh := time.NewTimer(2 * time.Second).C
+		select {
+		case <-ctx.Done():
+			return
+		case <-sleepCh:
+		}
+	}
+}
+
+func (e *Executor) tasksTimeoutCleaner(ctx context.Context) {
+	for _, rtID := range e.runningTasks.ids() {
+		rt, ok := e.runningTasks.get(rtID)
+		if !ok {
+			continue
+		}
+
+		rt.Lock()
+		if rt.et.Status.Phase == types.ExecutorTaskPhaseRunning && rt.et.Spec.TaskTimeoutInterval != 0 && time.Since(*rt.et.Status.StartTime) > rt.et.Spec.TaskTimeoutInterval {
+			rt.et.Status.Phase = types.ExecutorTaskPhaseTimedout
+			rt.et.Status.EndTime = util.TimeP(time.Now())
+
+			if rt.et.Status.SetupStep.Phase == types.ExecutorTaskPhaseRunning {
+				rt.et.Status.SetupStep.Phase = types.ExecutorTaskPhaseFailed
+				rt.et.Status.SetupStep.EndTime = util.TimeP(time.Now())
+			}
+			for _, s := range rt.et.Status.Steps {
+				if s.Phase == types.ExecutorTaskPhaseRunning {
+					s.Phase = types.ExecutorTaskPhaseFailed
+					s.EndTime = util.TimeP(time.Now())
+				}
+			}
+
+			if err := e.sendExecutorTaskStatus(ctx, rt.et); err != nil {
+				e.log.Err(err).Send()
+			}
```

Comment:
You cannot set the et.Status.Phase here since the task is not yet stopped. You should record that it was stopped due to timeout and then update the phase when really stopped. The unique place that should update the `rt.et.Status` is `executeTask`

Reply:
I meant to add a Timedout field to `runningTask`, then in `executeTask` when `et.Status.Phase` is updated you should check this value to choosed between failed or timedout.